Formula/solargraph.rb

```diff
@@ -0,0 +1,117 @@
+class Solargraph < Formula
+  desc "Ruby language server"
+  homepage "https://solargraph.org"
+  # Must be git, because solargraph.gemspec uses git ls-files
+  url "https://github.com/castwide/solargraph.git",
+      tag:      "v0.44.2",
+      revision: "e50a6dc4b43e2183e245aded8d23f0003d8c6bf7"
+  license "MIT"
+
+  uses_from_macos "ruby"
+
+  def install
+    ENV["GEM_HOME"] = libexec
+    system "gem", "build", "#{name}.gemspec"
+    system "gem", "install", "#{name}-#{version}.gem"
+    bin.install libexec/"bin/#{name}"
+    bin.env_script_all_files(libexec/"bin", GEM_HOME: ENV["GEM_HOME"])
+  end
+
+  test do
```

Comment:
Can this test not be run through a port instead of stdin? That would make it a lot easier. If that is not possible, see https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/kotlin-language-server.rb on how to use Homebrew standards for this

Reply:
@SMillerDev Thanks for the input! Can I have some more detail about the following so I can fix the issue?

> Can this test not be run through a port instead of stdin? That would make it a lot easier.

Yes, solargraph can run using a TCP socket with `solargraph socket --port=<port>`, but I'm not familiar enough with Ruby or Homebrew formulas to understand why using a socket/port is preferred over stdio.  I would appreciate it if you could explain this for me. 

The reason I decided to use stdio is because most language servers implement the stdio communication channel (all that I am aware of), but not all servers implement sockets (e.g. sorbet, bash-language-server, etc). I wanted to write a basic initialization-shutdown test that could be used as a template in multiple formulae for tools that implement the language server protocol, so stdio seemed to be better.

> If that is not possible, see https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/kotlin-language-server.rb on how to use Homebrew standards for this

I definitely want to follow the standards, so can you point me to which part of the test is non-standard? Is it the use of `Open3.popen3` instead of `pipe_output`?

Reply:
> Yes, solargraph can run using a TCP socket with solargraph socket --port=<port>, but I'm not familiar enough with Ruby or Homebrew formulas to understand why using a socket/port is preferred over stdio. I would appreciate it if you could explain this for me.

It makes the tests much easier. You start the server and do a curl request. That's all it needs to do in that case. The fact that your test defines functions is a strong indicator to me that it's waaaay too complex.

> I definitely want to follow the standards, so can you point me to which part of the test is non-standard? Is it the use of Open3.popen3 instead of pipe_output?

Yes, that is the main difference. You made a perfectly fine functional test for the project in Ruby. It just needs some tweaking to be a good formula test in Homebrew

Reply:
So since it seems like complexity is one of the main issues, I simplified the test to just do the initialization request for the language server. Personally I was a little worried about the language server not properly cleaning up resources if forcibly shutdown without the shutdown request and exit notification, but I don't think there really is a way to do the proper sequence while keeping the test concise.

> You start the server and do a curl request.

I looked into this method using curl, but I couldn't figure out how to do it. The language server listens for raw JSON RPC messages on a plain TCP socket. It doesn't support JSON RPC over HTTP. I guess I could do something like:

```ruby
require "socket"

port = free_port
fork do
  exec "#{bin}/solargraph", "socket", "--port=#{port}"
end

sleep 3 # wait for server to start up
socket = TCPSocket.open "127.0.0.1", port

socket.write "Content-Length: ..."
assert_match(/^Content-Length: \d+/i, socket.readline)
```

but this seems more complex than using stdio.

Lastly, I tried to write this test using `pipe_output` as requested, but it does not work with solargraph. The reason is that the language server shuts down before anything is written to stdout. It takes some time for the language server to asynchronously process the initialization request, and output anything to stdout. `pipe_output` doesn't allow for this so I am still using the `Open3.popen3` method as recommended in the [documentation](https://rubydoc.brew.sh/Formula#test-class_method). This allows me to call `sleep` so that the server has time to output the response to stdout.